do not modify CMAKE_BUILD_TYPE to not confuse IDE

codingdave opened the pull request:
(This is the same but fixed PR as https://github.com/darktable-org/rawspeed/pull/221)

Using Visual Studio Code with the cmake extension does not work with rawspeed. The reason is that the BUILD_TYPE is changed and such the cmake extension gets confused. You will find this error in the console:

[cms-driver] Found no matching code model for the current build type. This shouldn't be possible

See this issue here: microsoft/vscode-cmake-tools#774

This patch changes the file build-type.cmake to not modify the BUILD_TYPE but instead do its sanity checks on another copy of it. I hope it does not break the current logic, at least it makes rawspeed work with VS Code and the cmake extension work flawlessly.

Kind regards
David

codingdave committed:
do not modify CMAKE_BUILD_TYPE to not confuse IDE

LebedevRI commented:
So TLDR is that the cmakefiles must not modify `CMAKE_BUILD_TYPE`, but keep it
exactly as it was set? If that is the only thing i suspect it could be okay, but then you also need to
fix every place where cmakelists relied on the canonicalization you are removing.

codingdave committed:
changes RELEASE to Release and RELWITHDEBINFO to RelWithDebInfo

LebedevRI commented on cmake/compiler-flags.cmake:
```diff
@@ -23,10 +23,10 @@ set(CMAKE_CXX_EXTENSIONS OFF)
 include(debug-info)
 
-if(CMAKE_BUILD_TYPE STREQUAL "RELEASE")
+if(CMAKE_BUILD_TYPE STREQUAL "Release")
```
Note that this is not correct fix, this will now errneously not match `rELEASE` string

codingdave replied:
ok, I will check tomorrow. Thanks for the feedback.

codingdave replied:
Hmm, how come you say `rELEASE` is a valid input? Looking at https://cmake.org/cmake/help/v3.0/variable/CMAKE_BUILD_TYPE.html:
> Possible values are empty, Debug, Release, RelWithDebInfo and MinSizeRel.

LebedevRI replied:
case-insensitivity is how it is currently handled, it should stay that way be it either via
the LHS of the diff, or by fixing the every CMAKE_BUILD_TYPE-dependent check
to be case-insensitive.

codingdave replied:
ok. please read my last commit message. If you persist on making it case-insensitive I will see how I can achieve that goal.

codingdave commented:
> So TLDR is that the cmakefiles must not modify CMAKE_BUILD_TYPE, but keep it
exactly as it was set?

Yes, that is correct. 

> If that is the only thing i suspect it could be okay, but then you also need to
fix every place where cmakelists relied on the canonicalization you are removing.

With the last commit I have been only looking at Release, Debug, and RelWithDebInfo. I did not change COVERAGE, SANITIZE, TSAN, or FUZZ.

codingdave committed:
removes CMAKE_BUILD_TYPE_UPPERCASE for 2 reasons:

1. It currently does not work since the CMAKE_BUILD_TYPE_UPPERCASE
   contains Release, Debug, RelWithDebInfo, ... which are not uppercase.
   The CMAKE_BUILD_TYPE will be made uppercase and then compared against
   this list, which obviously already will fail for Release !=
   RELEASE...
2. Since
   https://cmake.org/cmake/help/v3.0/variable/CMAKE_BUILD_TYPE.html
   writes:
   > Possible values are empty, Debug, Release, RelWithDebInfo and MinSizeRel.
   I don't see any necessity to do any upper/lowercasing anymore. If the
   CMAKE_BUILD_TYPE is not known, the user gets a pretty descriptive
   message why it fails and what BUILD_TYPE is supported:
   > "Unknown build type: ${CMAKE_BUILD_TYPE}. Please specify one of: ${CMAKE_CONFIGURATION_TYPES}"

codingdave commented:
Please read the commit message from: removes CMAKE_BUILD_TYPE_UPPERCASE for 2 reasons: f7fafb5

codingdave commented:
Why are travis and appveyor failing? Will they pass when my patches are taking care of the case-insensitivity?

LebedevRI commented:
Anything else?

codingdave commented:
Thank you @LebedevRI, I can confirm that the issue has been fixed. 

Kind regards
David
